cpu/stm32f1/periph/rtc.c

```diff
@@ -0,0 +1,267 @@
+    RTC->CNTL = counter_val & 0x0000ffff;
+    _rtc_exit_config_mode();
+}
+
+void rtc_init(void)
+{
+    /* save current time if RTC already works */
+    bool is_rtc_enable = _is_rtc_enable();
+    time_t cur_time = 0;
+    if (is_rtc_enable) {
+        cur_time = _rtc_get_time();
+    }
+
+    /* config RTC */
+    _rtc_config();
+
+    /* restore current time after config */
+    if (is_rtc_enable) {
+        _rtc_set_time(cur_time);
+    }
+}
+
+int rtc_set_time(struct tm *time)
+{
+    time_t timestamp = mktime(time);
+
+    if (timestamp == -1) {
+        return -1;
+    }
+
+    _rtc_set_time(timestamp);
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+int rtc_get_time(struct tm *time)
+{
+    time_t timestamp = _rtc_get_time();
+    localtime_r(&timestamp, time);
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+static void _rtc_enable_alarm(void)
+{
+    /* clear alarm flag */
+    RTC->CRL &= ~RTC_CRL_ALRF;
+
+    _rtc_enter_config_mode();
+    RTC->CRH |= (RTC_CRH_ALRIE);
+    _rtc_exit_config_mode();
+}
+
+static void _rtc_disable_alarm(void)
+{
+    _rtc_enter_config_mode();
+    RTC->CRH &= ~(RTC_CRH_ALRIE | RTC_CRH_SECIE);
+    _rtc_exit_config_mode();
+}
+
+static time_t _rtc_get_alarm_time(void)
+{
+    return (RTC->ALRH << 16) | RTC->ALRL;
+}
+
+static void _rtc_set_alarm_time(time_t alarm_time)
+{
+    _rtc_enter_config_mode();
+    RTC->ALRL = (alarm_time & 0x0000ffff);
+    RTC->ALRH = (alarm_time & 0xffff0000) >> 16;
+    _rtc_exit_config_mode();
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
+{
+    /* enable global RTC interrupt */
+    if (cb && !NVIC_GetEnableIRQ(RTC_IRQn)) {
+        NVIC_EnableIRQ(RTC_IRQn);
+        DEBUG("%s enable RTC IRQ\n", __func__);
+    }
+
+    /* save callback and argument */
+    isr_ctx.cb = cb;
+    isr_ctx.arg = arg;
+
+    /* set wakeup time */
+    time_t timestamp = mktime(time);
+
+    if (timestamp == -1) {
+        return -2;
+    }
+
+    _rtc_disable_alarm();
+    _rtc_set_alarm_time(timestamp);
+
+    /* enable Alarm */
+    _rtc_enable_alarm();
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+int rtc_get_alarm(struct tm *time)
+{
+    time_t timestamp = _rtc_get_alarm_time();
+    localtime_r(&timestamp, time);
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+static void _rtc_clear_isr_ctx(void)
+{
+    isr_ctx.cb = NULL;
+    isr_ctx.arg = NULL;
+}
+
+void rtc_clear_alarm(void)
+{
+    _rtc_disable_alarm();
+    _rtc_set_alarm_time(0);
+
+    _rtc_clear_isr_ctx();
+}
+
+void rtc_poweron(void)
+{
+    /* RTC is always on */
+    return;
+}
+
+void rtc_poweroff(void)
+{
+    /* RTC is always on */
+    return;
+}
+
+void isr_rtc(void)
+{
+    struct alarm_isr_ctx_t ctx = isr_ctx;
+
+    _rtc_disable_alarm();
+    _rtc_clear_isr_ctx();
```

Comment:
Huh? Why is this needed here?
Now the callback will never be called.

Reply:
I check this code before commit. Callback called fine. See copy `ctx = isr_ctx`. In `ctx` saved copy callback data.

Reply:
Oh, didn't notice.
But this seems entirely unnecessary. Why do you need to delete the callback?
Disabling the alarm ought to be enough.